File: linked_list/138_copy_list_with_random_pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        marker = head
        hset = {}
        
        while marker:
            hset[marker] = Node(marker.val)
            marker = marker.next

        marker = head
        while marker:
            hset[marker].next = hset.get(marker.next)
            hset[marker].random = hset.get(marker.random)
            marker = marker.next
            
        return hset[head]
        

File: linked_list/138_copy_list_with_random_pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    Map<Node, Node> hashMap = new HashMap<>();
    
    public Node copyRandomList(Node head) {
        if (head == null) return head;
        if (hashMap.containsKey(head)) return hashMap.get(head);
        
        Node temp = new Node(head.val, null, null);
        hashMap.put(head, temp);
        
        temp.next = this.copyRandomList(head.next);
        temp.random = this.copyRandomList(head.random);
        return temp;
    }
}